### PR TITLE
OpenEphysBinaryRawIO: Fixing ttl multichan

### PR DESCRIPTION
**Pull request by vigji**
Addressing #1437.

Currently, all the logic in the events parser for OpenEphys data assumes that there is a single stream of events (simply positive or negative `states`), and it breaks down when acquiring multiple digital signals at once (`states` having different absolute values for different channels).

This is a simple fix that looks for rising and falling edges separately for each of the digital channels.

Add simple testing relying on existing test data, which fails before the fix and passes afterwards.

### CODE REVIEW

**Comment by alejoe91** [comment]
Thanks @vigji! Looking great!

**Comment by vigji** [comment]
Great, let me know if there's anything needed before merging!

File: neo/rawio/openephysbinaryrawio.py
Comment (zm711) [inline neo/rawio/openephysbinaryrawio.py:267]: Could we get a slightly more explanatory comment (maybe above L266). Based on the code it seems like the rising are encode by +1 channel so if I'm at channel 1 then +1 is rising and -1 is falling and then for channel 2 it would be +2 and -2.  What would the array of states actually look like? So I'm imagining something like:

```
all channels are packed into the same `states` array. So the states array includes positive and negative values for each channel for example channel one rising would be +1 and channel one falling would be -1. So we can check for where each channel appears to get each TTL rise and fall for each channel
```

Or something like this. Since we have to support multiple versions it is helpful in the cases where the format is changed for us to know exactly how the former format was set up.

**Review by zm711 — COMMENTED** [review]
This looks good to me, although I would ask for one additional comment in the code to explain what version this works for (all or v0.6+ whatever) along with what the structure of the data is.

File: neo/rawio/openephysbinaryrawio.py
Comment (vigji) [reply neo/rawio/openephysbinaryrawio.py:272]: Ok, should have been fixed now!

**Review by zm711 — DISMISSED** [review]
This works for me, but Alessio knows this way better so I'll let him decide when to merge/do final review.

File: neo/rawio/openephysbinaryrawio.py
Comment (alejoe91) [inline neo/rawio/openephysbinaryrawio.py:284]: @vigji maybe it's better to skip the extend in case `rising_indices != falling_indices` here? Otherwise an issue with one channel will result in all durations being lost!

We could also add a warning in case this happens for a certain channel. What do you think?

**Review by alejoe91 — CHANGES_REQUESTED** [review]
See comment

File: neo/rawio/openephysbinaryrawio.py
Comment (vigji) [reply neo/rawio/openephysbinaryrawio.py:284]: I guess there is also a question of what the edge cases might be. Is there really a scenario where this can break down? The only option is the acquisition board to be missing rising/falling edges, but I am not sure at that point I would still use the data - it is possible that could have happened for other digital channels as well, and pairs of miss-detections could go unnoticed, in principle.

Have you seen or heard of any of those edge scenarios?

Comment (zm711) [reply neo/rawio/openephysbinaryrawio.py:284]: One case could be if I accidentally stop a recording midstimulus (a mouse wakes up or the head-fixation apparatus comes loose). So maybe I want to save data, but I don't have a fall for the last "channel" of stim.  So I agree with Alessio!

Comment (alejoe91) [reply neo/rawio/openephysbinaryrawio.py:284]: That would still work though because [here](https://github.com/iurillilab/python-neo/blob/fixing-ttl-multichan/neo/rawio/openephysbinaryrawio.py#L277-L281) we check if the first falling is before the first rising and viceversa.

Comment (vigji) [reply neo/rawio/openephysbinaryrawio.py:284]: you raise a valid point @zm711 , but addressing that would require completely changing the way the interface represents digital data (currently, timestamp+duration of the ON state). Not sure this is the place for that. I agree that the current representation is not the most general, and I would be in favour of changing it, but I assume there might be a lot of code by now depending on it (or not - the fact that the bug addressed in this PR was still around lets me wondering how many people use this interface for multi-digital channel experiment logging).

@alejoe91 so you suggest to just add a warning, with no changes to the current logic in the end?

Comment (zm711) [reply neo/rawio/openephysbinaryrawio.py:284]: He's saying add those lines on top of your PR to skip a channel if it doesn't fit those conditions.

And fair enough. We have to live with the code we have :)

Comment (vigji) [reply neo/rawio/openephysbinaryrawio.py:293]: Should be gtg now!

**Review by alejoe91 — APPROVED** [review]

**Comment by samuelgarcia** [comment]
Hi,
I was totally off for a while.
I read this very very quickly.
I suspect that this `np.where()` can slow down a lot the parse_header for long recording.
Is it a general or a corner case ?
We need to keep in mind that the parse header is done multiple time when multiprocessing on spikeinterface side so we need to be very carfull for this.
Do we have benchmark for this ?

**Comment by vigji** [comment]
Not data but rule of thumbs benchmark:
assuming 10 channels sampling TTL pulses at 100 Hz for 5 hours I generate a random array of integers from 0 to 10, `np.where()` then takes approx. 15 ms on my laptop (MacBook Pro M1 2022). This matches my experience that I never see long times even for fast digital signals recorded for a couple of hours.

Does this make sense?

**Comment by zm711** [comment]
@vigji do we have you in our author list? I'm prepping our documents for new release. Would you like me to add you to the list (and what is the institution you were at/or are at for your Neo contribution?)

**Comment by vigji** [comment]
I don't think so as it was my first contribution! 
Yes I'd like to, my institute is [optional: Center for Neuroscience and Cognitive systems], Istituto Italiano di Tecnologia (IIT), Italy

**Comment by zm711** [comment]
Perfect. I will add you before our next release :) Thanks for the contribution!

**Pull request merged**
